### ISSUE TITLE
StackOverflowException reporting error in overload resolution

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#24869]
Compile:
```
using System.Collections.Generic;
using System.Linq;
class C
{
    int? E;
    static void F(IEnumerable<C> c)
    {
        const int F =
        c.Select(o => new { E = o.E ?? F });
    }
}
```

**Comment by jcouv** [comment]
The repeating portion of the stack:
```C#
Microsoft.CodeAnalysis.DiagnosticInfo.GetArgumentsToUse(System.IFormatProvider formatProvider)	Unknown
Microsoft.CodeAnalysis.DiagnosticInfo.GetMessage(System.IFormatProvider formatProvider)	Unknown
Microsoft.CodeAnalysis.DiagnosticWithInfo.GetMessage(System.IFormatProvider formatProvider)	Unknown
Microsoft.CodeAnalysis.DiagnosticFormatter.Format(Microsoft.CodeAnalysis.Diagnostic diagnostic, System.IFormatProvider formatter)	Unknown
Microsoft.CodeAnalysis.CSharp.CSDiagnostic.ToString()	Unknown
Microsoft.CodeAnalysis.CSharp.UnboundLambdaState.CanonicallyCompareDiagnostics(Microsoft.CodeAnalysis.Diagnostic x, Microsoft.CodeAnalysis.Diagnostic y)	Unknown
Microsoft.CodeAnalysis.CSharp.FirstAmongEqualsSet<System.__Canon>.IsMoreCanonical(System.__Canon newItem, System.__Canon oldItem)	Unknown
Microsoft.CodeAnalysis.CSharp.FirstAmongEqualsSet<Microsoft.CodeAnalysis.Diagnostic>.IntersectWith(System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Diagnostic> items)	Unknown
Microsoft.CodeAnalysis.CSharp.UnboundLambdaState.GenerateSummaryErrors(Microsoft.CodeAnalysis.DiagnosticBag diagnostics)	Unknown
Microsoft.CodeAnalysis.CSharp.OverloadResolutionResult<System.__Canon>.HadLambdaConversionError(Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments arguments)	Unknown
Microsoft.CodeAnalysis.CSharp.OverloadResolutionResult<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol>.ReportDiagnostics<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol>(Microsoft.CodeAnalysis.CSharp.Binder binder, Microsoft.CodeAnalysis.Location location, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, string name, Microsoft.CodeAnalysis.CSharp.BoundExpression receiver, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments arguments, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol> memberGroup, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol typeContainingConstructor, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol delegateTypeBeingInvoked, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode queryClause, bool isMethodGroupConversion)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpressionContinued(Microsoft.CodeAnalysis.SyntaxNode node, Microsoft.CodeAnalysis.SyntaxNode expression, string methodName, Microsoft.CodeAnalysis.CSharp.OverloadResolutionResult<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol> result, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments analyzedArguments, Microsoft.CodeAnalysis.CSharp.MethodGroup methodGroup, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol delegateTypeOpt, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode queryClause)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindMethodGroupInvocation(Microsoft.CodeAnalysis.SyntaxNode syntax, Microsoft.CodeAnalysis.SyntaxNode expression, string methodName, Microsoft.CodeAnalysis.CSharp.BoundMethodGroup methodGroup, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments analyzedArguments, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode queryClause, bool allowUnexpandedForm)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpression(Microsoft.CodeAnalysis.SyntaxNode node, Microsoft.CodeAnalysis.SyntaxNode expression, string methodName, Microsoft.CodeAnalysis.CSharp.BoundExpression boundExpression, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments analyzedArguments, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode queryClause, bool allowUnexpandedForm)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpression(Microsoft.CodeAnalysis.CSharp.Syntax.InvocationExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, bool invoked, bool indexed)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, bool invoked, bool indexed)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindPossibleArrayInitializer(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol destinationType, Microsoft.CodeAnalysis.CSharp.Binder.BindValueKind valueKind, Microsoft.CodeAnalysis.DiagnosticBag diagnostics)	Unknown
Microsoft.CodeAnalysis.CSharp.Binder.BindVariableOrAutoPropInitializer(Microsoft.CodeAnalysis.CSharp.Syntax.EqualsValueClauseSyntax initializerOpt, Microsoft.CodeAnalysis.RefKind refKind, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol varType, Microsoft.CodeAnalysis.DiagnosticBag diagnostics)	Unknown
Microsoft.CodeAnalysis.CSharp.Symbols.SourceLocalSymbol.LocalWithInitializer.MakeConstantTuple(Microsoft.CodeAnalysis.CSharp.Symbols.LocalSymbol inProgress, Microsoft.CodeAnalysis.CSharp.BoundExpression boundInitValue)	Unknown
Microsoft.CodeAnalysis.CSharp.Symbols.SourceLocalSymbol.LocalWithInitializer.GetConstantValue(Microsoft.CodeAnalysis.SyntaxNode node, Microsoft.CodeAnalysis.CSharp.Symbols.LocalSymbol inProgress, Microsoft.CodeAnalysis.DiagnosticBag diagnostics)	Unknown
Microsoft.CodeAnalysis.CSharp.Symbols.LocalSymbol.HasConstantValue.get()	Unknown
Microsoft.CodeAnalysis.CSharp.SymbolDisplayVisitor.VisitLocal(Microsoft.CodeAnalysis.ILocalSymbol symbol)	Unknown
Microsoft.CodeAnalysis.CSharp.Symbols.LocalSymbol.Accept(Microsoft.CodeAnalysis.SymbolVisitor visitor)	Unknown
Microsoft.CodeAnalysis.CSharp.SymbolDisplay.ToDisplayParts(Microsoft.CodeAnalysis.ISymbol symbol, Microsoft.CodeAnalysis.SemanticModel semanticModelOpt, int positionOpt, Microsoft.CodeAnalysis.SymbolDisplayFormat format, bool minimal)	Unknown
Microsoft.CodeAnalysis.CSharp.SymbolDisplay.ToDisplayParts(Microsoft.CodeAnalysis.ISymbol symbol, Microsoft.CodeAnalysis.SymbolDisplayFormat format)	Unknown
Microsoft.CodeAnalysis.CSharp.MessageProvider.GetErrorDisplayString(Microsoft.CodeAnalysis.ISymbol symbol)	Unknown
Microsoft.CodeAnalysis.DiagnosticInfo.GetArgumentsToUse(System.IFormatProvider formatProvider)	Unknown
``` 

I couldn't see the entry point because VS truncates the stack.

**Comment by jcouv** [comment]
Found the top of the stack (first call to compiler):
```
075ae19c 2dfe1bd9 Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.Bind(Microsoft.CodeAnalysis.CSharp.Binder, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, Microsoft.CodeAnalysis.DiagnosticBag)
075ae20c 2dfe15b3 Microsoft.CodeAnalysis.CSharp.MethodBodySemanticModel.Bind(Microsoft.CodeAnalysis.CSharp.Binder, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, Microsoft.CodeAnalysis.DiagnosticBag)
075ae23c 2dfd1f5d Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode)
075ae318 2dfd1a4d Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetLowerBoundNode(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode)
075ae350 2dfd015d Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetBoundNodes(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode ByRef, Microsoft.CodeAnalysis.CSharp.BoundNode ByRef, Microsoft.CodeAnalysis.CSharp.BoundNode ByRef, Microsoft.CodeAnalysis.CSharp.BoundNode ByRef)
075ae3c4 2ff4fe32 Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetSymbolInfoWorker(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, SymbolInfoOptions, System.Threading.CancellationToken)
075ae414 2ff436b9 Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetSymbolInfoWorker(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, SymbolInfoOptions, System.Threading.CancellationToken)
075ae5b4 2ff4161f Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSymbolInfo(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, System.Threading.CancellationToken)
075ae698 2ff40f8b Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetSymbolInfo(Microsoft.CodeAnalysis.SemanticModel, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax, System.Threading.CancellationToken)
```

**Comment by cston** [comment]
The same exception is hit with simple command line compile: `csc -t:library program.cs`

**Comment by jcouv** [comment]
Thanks. This works as a repro:
```C#

        [Fact]
        [WorkItem(24869, "https://github.com/dotnet/roslyn/issues/24869")]
        public void TestCrash()
        {
            string source = @"
using System.Collections.Generic;
using System.Linq;
class C
{
    int? E;
    static void F(IEnumerable<C> c)
    {
        const int F =
        c.Select(o => new { E = o.E ?? F });
    }
}";
            var comp = CreateStandardCompilation(source, references: new[] { LinqAssemblyRef });
            comp.VerifyDiagnostics();
        }
```

**Comment by jcouv** [comment]
Fix was merged for 15.7
